### PR TITLE
KAFKA-6397: Consumer should not block setting positions of unavailable partitions

### PR DESCRIPTION
**Pull request by hachikuji**
Prior to this patch, the consumer always blocks in poll() if there are any partitions which are awaiting their initial positions. This behavior was inconsistent with normal fetch behavior since we allow fetching on available partitions even if one or more of the assigned partitions becomes unavailable _after_ initial offset lookup. With this patch, the consumer will do offset resets asynchronously, which allows other partitions to make progress even if the initial positions for some partitions cannot be found.

I have added several new unit tests in `FetcherTest` and `KafkaConsumerTest` to verify the new behavior. One minor compatibility implication worth mentioning is apparent from the change I made in `DynamicBrokerReconfigurationTest`. Previously it was possible to assume that all partitions had a fetch position after `poll()` completed with a non-empty assignment. This assumption is no longer generally true, but you can force the positions to be updated using the `position()` API which still blocks indefinitely until a position is available.

Note that this this patch also removes the logic to cache committed offsets in `SubscriptionState` since it was no longer needed (the consumer's `committed()` API always does an offset lookup anyway). In addition to avoiding the complexity of maintaining the cache, this avoids wasteful offset lookups to refresh the cache when `commitAsync()` is used.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 1]
```diff
@@ -393,113 +341,113 @@ else if (error.exception() instanceof RetriableException)
...
-            }
-        }
-        if (!partitionsWithNoOffsets.isEmpty()) {
-            throw new NoOffsetForPartitionException(partitionsWithNoOffsets);
+            Long timestamp = offsetResetStrategyTimestamp(partition);
+            if (timestamp != null)
+                offsetResetTimestamps.put(partition, timestamp);
         }
+
+        resetOffsetsAsync(offsetResetTimestamps);
```
Comment (tedyu) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:377]: Is it possible that offsetResetTimestamps is empty ?

Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:377]: It is possible, but seems fine? We could even get rid of the empty check a couple lines above.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 2]
```diff
@@ -1159,10 +1159,16 @@ public void assign(Collection<TopicPartition> partitions) {
...
 
-        long now = time.milliseconds();
-        long pollTimeout = Math.min(coordinator.timeToNextPoll(now), timeout);
+        long nowMs = time.milliseconds();
+        long remainingTimeMs = Math.max(0, timeout - (nowMs - startMs));
+        long pollTimeout = Math.min(coordinator.timeToNextPoll(nowMs), remainingTimeMs);
+
+        // We do not want to be stuck blocking in poll if we are missing some positions
+        // since the offset lookup may be backing off after a failure
+        if (!hasAllFetchPositions && pollTimeout > retryBackoffMs)
```
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:1168]: `retryBackoffMs` is typically much lower than request timeout. Should `poll(largeValue)`  poll until request timeout to fetch offsets? Perhaps not, just want to make sure I understand.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 3]
```diff
@@ -1359,7 +1365,7 @@ public void seekToBeginning(Collection<TopicPartition> partitions) {
             Collection<TopicPartition> parts = partitions.size() == 0 ? this.subscriptions.assignedPartitions() : partitions;
             for (TopicPartition tp : parts) {
                 log.debug("Seeking to beginning of partition {}", tp);
-                subscriptions.needOffsetReset(tp, OffsetResetStrategy.EARLIEST);
+                subscriptions.requestOffsetReset(tp, OffsetResetStrategy.EARLIEST);
```
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:1368]: I thought this was changed because it actually requests something, but it is just setting a flag?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 4]
```diff
@@ -393,113 +341,113 @@ else if (error.exception() instanceof RetriableException)
...
 
         return offsetsByTimes;
     }
 
-    private Map<TopicPartition, OffsetData> retrieveOffsetsByTimes(
-            Map<TopicPartition, Long> timestampsToSearch, long timeout, boolean requireTimestamps) {
+    private ListOffsetResult fetchOffsetsByTimes(Map<TopicPartition, Long> timestampsToSearch,
+                                                 long timeout,
+                                                 boolean requireTimestamps) {
+        ListOffsetResult result = ListOffsetResult.empty();
```
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: `empty()` suggests an immutable empty result that is reused, while this is actually creating a new instance to populate later. Perhaps a different name?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 5]
```diff
@@ -753,21 +770,36 @@ private void handleListOffsetResponse(Map<TopicPartition, Long> timestampsToSear
...
+                partitionsToRetry.add(topicPartition);
             }
         }
-        if (!future.isDone())
-            future.complete(timestampOffsetMap);
+
+        if (!unauthorizedTopics.isEmpty())
+            future.raise(new TopicAuthorizationException(unauthorizedTopics));
+        else
+            future.complete(new ListOffsetResult(fetchedOffsets, partitionsToRetry));
```
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:790]: What do we do if one of the exceptions is not retriable (e.g UNSUPPORTED_FOR_MESSAGE_FORMAT)? It seems to be handled differently from `TOPIC_AUTHORIZATION_FAILED`, but I wasn't sure why.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 6]
```diff
@@ -753,21 +770,36 @@ private void handleListOffsetResponse(Map<TopicPartition, Long> timestampsToSear
...
+    private static class ListOffsetResult {
+        private final Map<TopicPartition, OffsetData> fetchedOffsets;
+        private final Set<TopicPartition> partitionsToRetry;
+
+        public ListOffsetResult(Map<TopicPartition, OffsetData> fetchedOffsets, Set<TopicPartition> partitionsNeedingRetry) {
+            this.fetchedOffsets = fetchedOffsets;
+            this.partitionsToRetry = partitionsNeedingRetry;
+        }
+
+        public static ListOffsetResult empty() {
```
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: Added a note earlier about renaming this since this is used as `newInstance` rather than `empty`.

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java [hunk 7]
```diff
@@ -117,14 +118,21 @@
...
+
     @Rule
     public ExpectedException expectedException = ExpectedException.none();
 
     @Test
     public void testConstructorClose() throws Exception {
         Properties props = new Properties();
         props.setProperty(ConsumerConfig.CLIENT_ID_CONFIG, "testConstructorClose");
-        props.setProperty(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, "some.invalid.hostname.foo.bar.local:9999");
+        props.setProperty(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, "invalid-23-8409-adsfsdj");
```
Comment (rajinisivaram) [inline clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:136]: Why was this change needed?

**Review by rajinisivaram — COMMENTED** [review]
@hachikuji Thanks for the PR. Looks good, left a few minor comments/questions. Will take another look through the unit tests tomorrow morning.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java (hunk 2 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:1168]: I'm trying to handle the case in which we are in the middle of a backoff after a failed offset fetch. There wouldn't be any pending IO in the worst case (say if we only had one partition assigned), so `poll()` would just block. We could alternatively do a bit more bookkeeping to keep track of when we are backing off, and when we have offset fetches pending, but it seemed simpler this way.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java (hunk 3 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:1368]: Yeah, I just didn't like the name. We had both `isOffsetResetNeeded` and `needOffsetReset`, which both sound like boolean flag checks. I wanted the mutation to be obvious in the name and I thought this was analogous with `Metadata.requestUpdate`. For some reason, we frown on names like `setOffsetResetNeeded`, but that would be an alternative which would arguably be less misleading.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java (hunk 5 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:790]: We treat `UNSUPPORTED_FOR_MESSAGE_FORMAT` as equivalent to not finding an offset for the timestamp being searched. It makes sense because timestamps didn't exist in the old format. I can clarify this in the comment.

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java (hunk 7 above)
Comment (hachikuji) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:136]: I was annoyed that it took 5 seconds for lookup of the previous invalid host to fail. Seemed like any value will work for the purpose of this test case so we can choose something which causes an immediate failure.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java (hunk 3 above)
Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:1368]: OK, since we have `Metadata.requestUpdate`, let's leave as is.

**Comment by rajinisivaram** [comment]
@hachikuji Looks like this needs a rebase since compilation of `KafkaConsumerTest` is failing.

**Comment by hachikuji** [comment]
@rajinisivaram Thanks for reviewing. I've rebased.

**Review by rajinisivaram — APPROVED** [review]
@hachikuji Thanks for the updates. LGTM, feel free to merge when builds complete.
